numpy/core/src/multiarray/ctors.c

```diff
@@ -3588,23 +3609,23 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
         goto done;
     }
 
-    ret = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type, dtype, 1,
-                                                &elcount, NULL,NULL, 0, NULL);
+    ret = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type, dtype, ndim,
+                                                elcount, NULL,NULL, 0, NULL);
     dtype = NULL;
     if (ret == NULL) {
         goto done;
     }
-    for (i = 0; (i < count || count == -1) &&
-             (value = PyIter_Next(iter)); i++) {
-        if (i >= elcount) {
+    for (i = 0; (i < count || count == -1) && value; i++) {
```

Comment:
Should we check and raise an error if `count < -1`? I know that wasn't done originally but it looks like an eversight.